File: ASCIIArt.WinConsole/ConsoleDisplayInfo.cs
using ASCIIArt.Engine;
using ASCIIArt.WinConsole.Native;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.Drawing.Text;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Threading;

namespace ASCIIArt.WinConsole
{
    public class ConsoleDisplayInfo : IConsoleDisplayInfo
    {
        public int WidthInColumns => Console.WindowWidth;
        public int HeightInRows => Console.WindowHeight;

        public int CharPixelWidth => CharPixelSize.Width;

        public int CharPixelHeight => CharPixelSize.Height;

        public Size ClientAreaSize =>
            new Size(WidthInColumns * CharPixelWidth, HeightInRows * CharPixelHeight);

        public Size CharPixelSize { get; protected set; }

        protected Lazy<IEnumerable<(string, byte[])>> _PrintableChars { get; set; }

        public IEnumerable<(string, byte[])> PrintableChars
        {
            get => _PrintableChars.Value;
        }


        public ConsoleDisplayInfo()
        {
            // Lock Console
            var hWindow = User32.GetForegroundWindow();
            var style = (WindowStyles)User32.GetWindowLong(hWindow, WindowLongFlags.GWL_STYLE);
            style &= ~WindowStyles.WS_SIZEBOX;
            User32.SetWindowLong(hWindow, WindowLongFlags.GWL_STYLE, (int)style);

            Initialize();
        }

        public void SetConsoleSize(in Size size)
        {
            SetConsoleSize(size.Width, size.Height);
        }

        public void SetConsoleSize(int width, int height)
        {
            Console.SetWindowSize(width, height);
            Initialize();
        }

        protected virtual void Initialize()
        {
            // Fetch Font Info
            var hStdOut = Kernel32.GetStdHandle(StdHandle.STD_OUTPUT_HANDLE);
            var fontInfo = CONSOLE_FONT_INFOEX.Create();
            Kernel32.GetCurrentConsoleFontEx(hStdOut, false, ref fontInfo);

            CharPixelSize = fontInfo.dwFontSize;

            _PrintableChars = new Lazy<IEnumerable<(string, byte[])>>(
                () => GetAvailableCharBitmaps(GetPrintableChars(fontInfo)));
        }

        private List<(string, byte[])> GetAvailableCharBitmaps(char[] printable)
        {

            var result = new List<(string, byte[])>();
            var currentWindow = Kernel32.GetConsoleWindow();
            using (var stream = new MemoryStream())
            {
                foreach (var chunck in printable.Chunk(WidthInColumns * HeightInRows))
                {
                    Console.Clear();
                    Console.Write(string.Concat(chunck));
                    //Wait for screen to update
                    Thread.Sleep(300);
                    using (var bitmap = CaptureScreen(currentWindow))
                    {
                        var i = 0;
                        foreach (var c in chunck)
                        {
                            var loc = new Point()
                            {
                                X = i % WidthInColumns * CharPixelWidth,
                                Y = i / WidthInColumns * CharPixelHeight,
                            };
                            var charRect = new Rectangle(loc, CharPixelSize);
                            var piece = bitmap.Clone(charRect, bitmap.PixelFormat);
                            stream.Seek(0, SeekOrigin.Begin);
                            stream.SetLength(0);
                            piece.Save(stream, ImageFormat.Png);
                            result.Add((c.ToString(), stream.ToArray()));
                            i++;
                        }
                    }
                }
            }
            Console.Clear();
            return result;
        }

        private static char[] GetPrintableChars(in CONSOLE_FONT_INFOEX consoleInfo)
        {
            var fontCollection = new InstalledFontCollection();
            try
            {
                var fam = new FontFamily(consoleInfo.FaceName, fontCollection);

                var font = new Font(fam, 20);
                var ranges = GetFontUnicodeRanges(font);
                return Enumerable
                    .Range(1, char.MaxValue)
                    .Where(c => ranges.Any(r => c >= r.Low && c < r.High))
                    .Select(Convert.ToChar)
                    .Where(c => !char.IsWhiteSpace(c) && !char.IsControl(c))
                    .Append(' ')
                    .ToArray();
            }
            catch(ArgumentException)
            {
                return Enumerable
                    .Range(1, 127)
                    .Select(Convert.ToChar)
                    //.Where(c => !char.IsWhiteSpace(c) && User32.VkKeyScanEx(c, layout) != -1)
                    .Where(c => !char.IsWhiteSpace(c) && !char.IsControl(c))
                    .Append(' ')
                    .ToArray();
            }

        }

        private static Bitmap CaptureScreen(IntPtr hWindow)
        {
            User32.GetClientRect(hWindow, out var rect);
            var result = new Bitmap(
                rect.Width,
                rect.Height,
                PixelFormat.Format32bppArgb);
            using (var bmpGraphics = Graphics.FromImage(result))
            {
                POINT loc = rect.Location;
                User32.ClientToScreen(hWindow, ref loc);
                bmpGraphics.CopyFromScreen(loc, Point.Empty, rect.Size);
            }
            return result;
        }

        private static List<FontRange> GetFontUnicodeRanges(Font font)
        {
            Graphics g = Graphics.FromHwnd(IntPtr.Zero);
            IntPtr hdc = g.GetHdc();
            IntPtr hFont = font.ToHfont();
            IntPtr old = GDI32.SelectObject(hdc, hFont);
            uint size = GDI32.GetFontUnicodeRanges(hdc, IntPtr.Zero);
            IntPtr glyphSet = Marshal.AllocHGlobal((int)size);
            try
            {
                GDI32.GetFontUnicodeRanges(hdc, glyphSet);
                List<FontRange> fontRanges = new List<FontRange>();
                int count = Marshal.ReadInt32(glyphSet, 12);
                for (int i = 0; i < count; i++)
                {
                    FontRange range = new FontRange();
                    range.Low = (ushort)Marshal.ReadInt16(glyphSet, 16 + i * 4);
                    range.High = (ushort)(range.Low + Marshal.ReadInt16(glyphSet, 18 + i * 4) - 1);
                    fontRanges.Add(range);
                }
                return fontRanges;
            }
            finally
            {
                GDI32.SelectObject(hdc, old);
                Marshal.FreeHGlobal(glyphSet);
                g.ReleaseHdc(hdc);
                g.Dispose();
            }
        }
    }

    internal static class LinqEx
    {
        public static IEnumerable<IEnumerable<T>> Chunk<T>(this IEnumerable<T> source, int chunksize)
        {
            if (chunksize <= 0)
            {
                throw new ArgumentException("Chunk size must be greater than zero.", nameof(chunksize));
            }
            while (source.Any())
            {
                yield return source.Take(chunksize);
                source = source.Skip(chunksize);
            }
        }
    }
}

File: ASCIIArt.WinConsole.Native/ConsoleDisplayInfo.cpp
#include "pch.h"

#include <memory>
#include <dwrite_1.h>
#include <vector>
#include "ConsoleDisplayInfo.h"

ASCIIArtWinConsole::Native::ConsoleDisplayInfo::ConsoleDisplayInfo():
    fontInfo(new CONSOLE_FONT_INFOEX{sizeof(CONSOLE_FONT_INFOEX)})
{
    // Lock Console
    auto hWindow { GetForegroundWindow() };
    auto style { GetWindowLong(hWindow, GWL_STYLE) };
    style &= ~WS_SIZEBOX;
    SetWindowLong(hWindow, GWL_STYLE, style);
    _PrintableChars = gcnew Lazy<IReadOnlyDictionary<String^, array<byte>^>^>(
        gcnew Func<IReadOnlyDictionary<String^, array<byte>^>^>(this, &ConsoleDisplayInfo::GetAvailableCharBitmaps)
        );
    Initialize();
}

void ASCIIArtWinConsole::Native::ConsoleDisplayInfo::SetConsoleSize(int width, int height)
{
    Console::SetWindowSize(width, height);
    Initialize();
}

ASCIIArtWinConsole::Native::ConsoleDisplayInfo::~ConsoleDisplayInfo()
{
    this->!ConsoleDisplayInfo();
}

ASCIIArtWinConsole::Native::ConsoleDisplayInfo::!ConsoleDisplayInfo()
{
    delete fontInfo;
    fontInfo = nullptr;
}

void ASCIIArtWinConsole::Native::ConsoleDisplayInfo::Initialize()
{
    auto hStdOut { GetStdHandle(STD_OUTPUT_HANDLE) };
    GetCurrentConsoleFontEx(hStdOut, FALSE, fontInfo);
    _CharPixelSize = Size(fontInfo->dwFontSize.X, fontInfo->dwFontSize.Y);
}

IReadOnlyDictionary<String^, array<byte>^>^ ASCIIArtWinConsole::Native::ConsoleDisplayInfo::GetAvailableCharBitmaps()
{
    using namespace System::IO;
    using namespace System::Threading;
    using namespace System::Drawing;
    using namespace System::Drawing::Imaging;
    using namespace System::Collections::Generic;

    auto printable = GetPrintableChars();
    auto result{ gcnew Dictionary<String^, array<byte>^>() };
    auto currentWindow{ GetConsoleWindow() };
    auto stream{ gcnew MemoryStream() };
    auto chunkSize{ WidthInColumns * HeightInRows };

    auto buffer{ gcnew List<Tuple<Char, Point>^>() };
    auto e = printable->GetEnumerator();
    bool next = true;
    for (decltype(chunkSize) begin; next; begin += chunkSize)
    {
        Console::Clear();
        buffer->Clear();
        for (decltype(begin) i = 0; i < begin + chunkSize && (next = e->MoveNext()); i++)
        {
            Point loc(i % WidthInColumns * CharPixelSize.Width, i / WidthInColumns * CharPixelSize.Height);
            auto c = e->Current;
            Console::Write(c);
            buffer->Add(gcnew Tuple<Char, Point>(c, loc));
        }
        Thread::Sleep(300);
        auto bitmap{ CaptureScreen(currentWindow) };
        for each(auto pair in buffer)
        {      
            auto c = pair->Item1;
            auto loc = pair->Item2;
            System::Drawing::Rectangle charRect(loc, CharPixelSize);
            auto piece = bitmap->Clone(charRect, bitmap->PixelFormat);
            stream->Seek(0, SeekOrigin::Begin);
            stream->SetLength(0);
            piece->Save(stream, ImageFormat::Png);
            result->Add(c.ToString(), stream->ToArray());
        }
    }
    Console::Clear();
    return result;
}

[StructLayout(LayoutKind::Sequential)]
value struct Range
{
    UINT32 first;
    UINT32 last;
};

ref class UnicodeRangeEmrt :
    public IEnumerator<Char>, public IDisposable
{
    IList<Range>^ ranges;
    Char _Current;
    std::size_t iRange;
public:
    explicit UnicodeRangeEmrt(IList<Range>^ ranges):
        ranges(ranges), _Current(), iRange()
    {

    }
    ~UnicodeRangeEmrt()
    {

    }

    virtual bool MoveNext()
    {
        if (iRange == ranges->Count)
        {
            return false;
        }
        if (_Current++ < ranges[iRange].last)
        {
            return true;
        }
        else if (iRange++ < ranges->Count)
        {
            _Current = ranges[iRange].first;
            return true;
        }
        else
        {
            return false;
        }
    }
    virtual void Reset()
    {
        throw gcnew NotSupportedException();
    }
    virtual property Char Current
    {
        Char get()
        {
            return _Current;
        }
    }

protected:
    virtual property System::Object^ NonGenericCurrent
    {
        System::Object^ get() = System::Collections::IEnumerator::Current::get
        {
            return Current;
        }
    }
};

ref class UnicodeRangeEmrb :
    public IEnumerable<Char>
{
    IList<Range>^ ranges;
public:
    explicit UnicodeRangeEmrb(IList<Range>^ ranges) :
        ranges(ranges)
    {

    }

    virtual System::Collections::Generic::IEnumerator<wchar_t>^ GetEnumerator()
    {
        return gcnew UnicodeRangeEmrt(ranges);
    }

protected:
    virtual System::Collections::IEnumerator^ NonGenericGetEnumerator() =
        System::Collections::IEnumerable::GetEnumerator
    {
        return gcnew UnicodeRangeEmrt(ranges);
    }
};

IEnumerable<Char>^ ASCIIArtWinConsole::Native::ConsoleDisplayInfo::GetPrintableChars()
{
    CoInitialize(nullptr);
    try
    {
        CComPtr<IDWriteFactory1> factory;
        HRESULT hr;
        hr = DWriteCreateFactory(
            DWRITE_FACTORY_TYPE_SHARED,
            __uuidof(IDWriteFactory1),
            (IUnknown**)(&factory));
        Check(hr);
        CComPtr<IDWriteFontCollection> fontCollection;
        hr = factory->GetSystemFontCollection(&fontCollection);
        Check(hr);
        CComPtr<IDWriteFontFamily> fam;
        BOOL exist;
        UINT index;
        hr = fontCollection->FindFamilyName(fontInfo->FaceName, &index, &exist);
        Check(hr);
        if (!exist)
        {
            //TODO: Load terminal font
            throw gcnew NotImplementedException();
        }
        hr = fontCollection->GetFontFamily(index, &fam);
        CComPtr<IDWriteFont> _font;
        hr = fam->GetFont(0, &_font);
        Check(hr);
        CComPtr<IDWriteFont1> font;
        hr = _font->QueryInterface<IDWriteFont1>(&font);
        Check(hr);
        UINT32 rangeCount;
        hr = font->GetUnicodeRanges(0, nullptr, &rangeCount);
        if (hr != E_NOT_SUFFICIENT_BUFFER)
        {
            Check(hr);
        }
        std::vector<DWRITE_UNICODE_RANGE> ranges(rangeCount);
        hr = font->GetUnicodeRanges(ranges.size(), ranges.data(), &rangeCount);
        Check(hr);

        auto rangeBuffer{ gcnew List<Range>(ranges.size()) };        
        for (auto& r : ranges)
        {
            auto p = IntPtr(&r);
            rangeBuffer->Add(Marshal::PtrToStructure<Range>(p));
        }
        return gcnew UnicodeRangeEmrb(rangeBuffer);
    }
    finally
    {
        CoUninitialize();
    }
}

Bitmap^ ASCIIArtWinConsole::Native::ConsoleDisplayInfo::CaptureScreen(HWND hWindow)
{
    using namespace System::Drawing::Imaging;
    RECT rect;
    GetClientRect(hWindow, &rect);
    auto result{
        gcnew Bitmap(
        rect.right - rect.left,
        rect.top - rect.bottom,
        PixelFormat::Format32bppArgb) 
    };
    auto bmpGraphics{ Graphics::FromImage(result) };
    POINT _loc{ rect.left, rect.top };
    ClientToScreen(hWindow, &_loc);
    Point loc(_loc.x, _loc.y);
    Size size(rect.right - rect.left, rect.top - rect.bottom);
    bmpGraphics->CopyFromScreen(loc, Point::Empty, size);
    return result;
}

void ASCIIArtWinConsole::Native::ConsoleDisplayInfo::Check(HRESULT hr)
{
    Marshal::ThrowExceptionForHR(hr);
}

int ASCIIArtWinConsole::Native::ConsoleDisplayInfo::HeightInRows::get()
{
    return Console::WindowHeight;
}

int ASCIIArtWinConsole::Native::ConsoleDisplayInfo::WidthInColumns::get()
{
    return Console::WindowWidth;
}

Size ASCIIArtWinConsole::Native::ConsoleDisplayInfo::ClientAreaSize::get()
{
    return Size(WidthInColumns * CharPixelSize.Width, HeightInRows * CharPixelSize.Height);
}

Size ASCIIArtWinConsole::Native::ConsoleDisplayInfo::CharPixelSize::get()
{
    return this->_CharPixelSize;
}

IReadOnlyDictionary<String^, array<byte>^>^ ASCIIArtWinConsole::Native::ConsoleDisplayInfo::PrintableChars::get()
{
    return this->_PrintableChars->Value;
}